src/mpmc_queue.hpp

```diff
@@ -120,9 +117,6 @@ class MPMCQueue : public Allocated {
         pos = head_.load(MEMORY_ORDER_RELAXED);
       }
     }
-
-    // never taken
-    return false;
```

Comment:
These two seemed a bit more perplexing to me.  I'm _assuming_ they exist to make code analysis happy but I'm wondering if maybe it isn't better to do something like the following:

```patch
diff --git a/src/mpmc_queue.hpp b/src/mpmc_queue.hpp
index f6cd146c..53abdf5d 100644
--- a/src/mpmc_queue.hpp
+++ b/src/mpmc_queue.hpp
@@ -69,11 +69,7 @@ public:
         // weak compare is faster, but can return spurious results
         // which in this instance is OK, because it's in the loop
         if (tail_.compare_exchange_weak(pos, pos + 1, MEMORY_ORDER_RELAXED)) {
-          // set the data
-          node->data = data;
-          // increment the sequence so that the tail knows it's accessible
-          node->seq.store(pos + 1, MEMORY_ORDER_RELEASE);
-          return true;
+          break;
         }
       } else if (dif < 0) {
         // if seq is less than head seq then it means this slot is
@@ -85,8 +81,11 @@ public:
       }
     }
 
-    // never taken
-    return false;
+    // set the data
+    node->data = data;
+    // increment the sequence so that the tail knows it's accessible
+    node->seq.store(pos + 1, MEMORY_ORDER_RELEASE);
+    return true;
   }
 
   bool dequeue(T& data) {
@@ -104,12 +103,7 @@ public:                                                                                                                                                                                          
         // weak compare is faster, but can return spurious results                                                                                                                                                   
         // which in this instance is OK, because it's in the loop                                                                                                                                                    
         if (head_.compare_exchange_weak(pos, pos + 1, MEMORY_ORDER_RELAXED)) {                                                                                                                                       
-          // set the output                                                                                                                                                                                          
-          data = node->data;                                                                                                                                                                                         
-          // set the sequence to what the head sequence should be next                                                                                                                                               
-          // time around                                                                                                                                                                                             
-          node->seq.store(pos + mask_ + 1, MEMORY_ORDER_RELEASE);                                                                                                                                                    
-          return true;                                                                                                                                                                                               
+          break;                                                                                                                                                                                                     
         }                                                                                                                                                                                                            
       } else if (dif < 0) {                                                                                                                                                                                          
         // if seq is less than head seq then it means this slot is                                                                                                                                                   
@@ -121,8 +115,12 @@ public:                                                                                                                                                                                          
       }                                                                                                                                                                                                              
     }                                                                                                                                                                                                                
                                                                                                                                                                                                                      
-    // never taken                                                                                                                                                                                                   
-    return false;                                                                                                                                                                                                    
+    // set the output                                                                                                                                                                                                
+    data = node->data;                                                                                                                                                                                               
+    // set the sequence to what the head sequence should be next                                                                                                                                                     
+    // time around                                                                                                                                                                                                   
+    node->seq.store(pos + mask_ + 1, MEMORY_ORDER_RELEASE);                                                                                                                                                          
+    return true;                                                                                                                                                                                                     
   }                                                                                                                                                                                                                  
                                                                                                                                                                                                                      
   bool is_empty() const { 
```

This change should be logically equivalent to the previous impl.  It also brings us closer to the [original implementation](https://www.1024cores.net/home/lock-free-algorithms/queues/bounded-mpmc-queue).

Whaddya think @SeverinLeonhardt ?

Reply:
My change is definitely to make code analysis happy, so maybe either way some compiler will complain.

Your suggested changes look good to me and nicely work around it. Actually I'd say this code is even better structured. If I understand GitHub right you should be able to push commits to the PR. Let me know if that doesn't work, then I'll add it.

Reply:
If I'm reading [these docs](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request) right that's something that can be setup when the PR is created.  Honestly it's probably faster/easier for you to incorporate the changes directly in another commit @SeverinLeonhardt.  If that works for you let's just go with that; I'll take another look once the changes are in.

Reply:
@absurdfarce Sorry for the delay, commit is added to the PR now. I've also used the patched version to run some tests locally and it seemed to all work fine.